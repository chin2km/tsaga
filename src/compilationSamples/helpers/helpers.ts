import * as path from 'path';
// tslint:disable-next-line:no-implicit-dependencies (Just used in tests)
import * as ts from 'typescript';

interface ReducedDiagnostic {
  message: string;
  fileName?: string;
  line?: number;
  character?: number;
}

export const defaultCompilerOptions: ts.CompilerOptions = {
  noEmit: true,
  strict: true,
  target: ts.ScriptTarget.ES2016,
  module: ts.ModuleKind.CommonJS,
  jsxFactory: 'react',
};

export function reducedDiagnostic(
  diagnostic: ts.Diagnostic,
): ReducedDiagnostic {
  if (diagnostic.file) {
    const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(
      diagnostic.start!,
    );
    const message = ts.flattenDiagnosticMessageText(
      diagnostic.messageText,
      '\n',
    );

    return {
      message,
      fileName: path.basename(diagnostic.file.fileName),
      line,
      character,
    };
  } else {
    return {
      message: ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n'),
    };
  }
}

export function getDiagnostics(
  fileNames: string[],
  options: ts.CompilerOptions,
): ReducedDiagnostic[] {
  const program = ts.createProgram(fileNames, options);
  const emitResult = program.emit();

  const diagnostics = ts
    .getPreEmitDiagnostics(program)
    .concat(emitResult.diagnostics);

  return diagnostics.map(reducedDiagnostic);
}

export function diagnosticsForFile(fileName: string) {
  return getDiagnostics([fileName], defaultCompilerOptions);
}
