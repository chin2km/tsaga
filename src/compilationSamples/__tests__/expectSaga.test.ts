import { diagnosticsForFile } from '../helpers/helpers';

test('expectSaga forces compiler errors when invalid types are used', async () => {
  const diagnostics = diagnosticsForFile(`${__dirname}/../samples/expectSaga-test.ts`);

  expect(diagnostics).toMatchSnapshot();

  // explicit alignment with comments in file, to catch accidental snapshot overwrite
  expect(diagnostics.length).toBe(4);
});
