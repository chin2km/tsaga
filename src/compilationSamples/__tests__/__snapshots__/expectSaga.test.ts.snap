// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expectSaga forces compiler errors when invalid types are used 1`] = `
Array [
  Object {
    "character": 27,
    "fileName": "expectSaga-test.ts",
    "line": 42,
    "message": "Argument of type '(state: { text: string; }) => string' is not assignable to parameter of type '(state: CountReducerState) => string'.
  Types of parameters 'state' and 'state' are incompatible.
    Property 'text' is missing in type 'CountReducerState' but required in type '{ text: string; }'.",
  },
  Object {
    "character": 53,
    "fileName": "expectSaga-test.ts",
    "line": 55,
    "message": "Argument of type '5' is not assignable to parameter of type 'boolean'.",
  },
  Object {
    "character": 20,
    "fileName": "expectSaga-test.ts",
    "line": 56,
    "message": "Argument of type 'Action<{ notText: string; }>' is not assignable to parameter of type 'Action<{ text: string; }>'.
  Property 'text' is missing in type '{ notText: string; }' but required in type '{ text: string; }'.",
  },
  Object {
    "character": 24,
    "fileName": "expectSaga-test.ts",
    "line": 65,
    "message": "Type 'string' is not assignable to type 'number'.",
  },
]
`;
